extensions/amp-a4a/0.1/amp-a4a.js

```diff
@@ -0,0 +1,685 @@
+    };
+    metaData.bodyUtf16CharOffsets = metaDataObj['bodyUtf16CharOffsets'];
+    if (!isValidOffsetArray(metaData.bodyUtf16CharOffsets)) {
+      // Invalid/Missing body offsets array.
+      throw new Error('Invalid/missing body offsets');
+    }
+    if (metaDataObj['cssUtf16CharOffsets']) {
+      metaData.cssUtf16CharOffsets = metaDataObj['cssUtf16CharOffsets'];
+      if (!isValidOffsetArray(metaData.cssUtf16CharOffsets)) {
+        throw new Error('Invalid CSS offsets');
+      }
+    }
+    // Validate array of two member number arrays
+    if (metaDataObj['cssReplacementRanges']) {
+      metaData.cssReplacementRanges = metaDataObj['cssReplacementRanges'];
+      if (!isArray(metaData.cssReplacementRanges)) {
+        throw new Error('Invalid CSS replacement ranges');
+      }
+      for (let i = 0; i < metaData.cssReplacementRanges.length; i++) {
+        if (!isValidOffsetArray(metaData.cssReplacementRanges[i])) {
+          throw new Error('Invalid CSS replacement ranges');
+        }
+      }
+    }
+    if (metaDataObj['bodyAttributes']) {
+      metaData.bodyAttributes = metaDataObj['bodyAttributes'];
+      if (typeof metaData.bodyAttributes !== 'string') {
+        throw new Error('Invalid body attributes');
+      }
+    }
+    if (metaDataObj['customElementExtensions']) {
+      metaData.customElementExtensions = metaDataObj['customElementExtensions'];
+      if (!isArray(metaData.customElementExtensions)) {
+        throw new Error('Invalid extensions');
+      }
+    }
+    if (metaDataObj['customStylesheets']) {
+      // Expect array of objects with at least one key being 'href' whose value
+      // is URL.
+      metaData.customStylesheets = metaDataObj['customStylesheets'];
+      if (!isArray(metaData.customStylesheets)) {
+        throw new Error('Invalid custom stylesheets');
+      }
+      for (let i = 0; i < metaData.customStylesheets.length; i++) {
+        const stylesheet = metaData.customStylesheets[i];
+        if (!isObject(stylesheet) || !stylesheet['href'] ||
+            typeof stylesheet['href'] !== 'string' ||
+            !/^https:\/\//i.test(stylesheet['href'])) {
+          throw new Error('Invalid custom stylesheets');
+        }
+      }
+    }
+    return metaData;
+  }
+
+  /**
+   * Extracts the body portion of the creative, according to directions in the
+   * metaData, and formats it for insertion into Shadow DOM.
+   * @param {string} creative from which CSS is extracted
+   * @param {!CreativeMetaDataDef} metaData Metadata object extracted from the
+   *    reserialized creative.
+   * @returns {string}  Body of AMP creative, surrounded by {@code
+   *     <amp-ad-body>} tags, and suitable for injection into Shadow DOM.
+   * @private
+   */
+  formatBody_(creative, metaData) {
+    const body = creative.substring(metaData.bodyUtf16CharOffsets[0],
+        metaData.bodyUtf16CharOffsets[1]);
+    let openString = '<' + AMP_BODY_STRING;
+    if (metaData.bodyAttributes) {
```

Comment:
Is the concern that we are adding free form string to the amp body element?  The validation service will already ensure that there is no javascript within this value and it determines the offsets from within the body tag so I'm not sure of the risk.  Please advise on how to setup the security review.  Thanks